bees/youtubebee/youtubefactory.go

```diff
@@ -0,0 +1,113 @@
+/*
+ *    Copyright (C) 2014-2017 Christian Muehlhaeuser
+ *
+ *    This program is free software: you can redistribute it and/or modify
+ *    it under the terms of the GNU Affero General Public License as published
+ *    by the Free Software Foundation, either version 3 of the License, or
+ *    (at your option) any later version.
+ *
+ *    This program is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    GNU Affero General Public License for more details.
+ *
+ *    You should have received a copy of the GNU Affero General Public License
+ *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *    Authors:
+ *      Christian Muehlhaeuser <muesli@gmail.com>
+ */
+
+package youtubebee
+
+import (
+	"github.com/muesli/beehive/bees"
+)
+
+// WebBeeFactory is a factory for WebBees.
+type YoutubeBeeFactory struct {
+	bees.BeeFactory
+}
+
+// New returns a new Bee instance configured with the supplied options.
+func (factory *YoutubeBeeFactory) New(name, description string, options bees.BeeOptions) bees.BeeInterface {
+	bee := YoutubeBee{
+		Bee: bees.NewBee(name, factory.ID(), description, options),
+	}
+	bee.ReloadOptions(options)
+
+	return &bee
+}
+
+// ID returns the ID of this Bee.
+func (factory *YoutubeBeeFactory) ID() string {
+	return "youtubebee"
+}
+
+// Name returns the name of this Bee.
+func (factory *YoutubeBeeFactory) Name() string {
+	return "Youtube"
+}
+
+// Description returns the description of this Bee.
+func (factory *YoutubeBeeFactory) Description() string {
+	return "HTTP Server that listens to a Youtube channel"
+}
+
+// Image returns the filename of an image for this Bee.
+func (factory *YoutubeBeeFactory) Image() string {
+	return factory.ID() + ".png"
+}
+
+// LogoColor returns the preferred logo background color (used by the admin interface).
+func (factory *YoutubeBeeFactory) LogoColor() string {
+	return "#ff0000"
+}
+
+// Options returns the options available to configure this Bee.
+func (factory *YoutubeBeeFactory) Options() []bees.BeeOptionDescriptor {
+	opts := []bees.BeeOptionDescriptor{
+		{
+			Name:        "address",
+			Description: "Which addr to listen on, eg: 0.0.0.0:12345",
+			Type:        "address",
+			Mandatory:   true,
+		},
+		{
+			Name:        "channel",
+			Description: "What is the link of the channel you want to receive push notifications for?",
+			Type:        "url",
+			Mandatory:   true,
+		},
+	}
+	return opts
+}
+
+// Events describes the available events provided by this Bee.
+func (factory *YoutubeBeeFactory) Events() []bees.EventDescriptor {
+	events := []bees.EventDescriptor{
+		{
+			Namespace:   factory.Name(),
+			Name:        "notification",
+			Description: "A push notification was sent by the Youtube channel",
+			Options: []bees.PlaceholderDescriptor{
```

Comment:
Is there any other information that can be extracted from YouTube's API? As it stands, this mostly just seems like a generic push, rather than something more actionable. If not, that's absolutely fine - I'm just wondering, so we can broaden it now if possible 😄

Reply:
Also, again, if we _can_ extract more information, then @muesli's suggestion of dividing the current monolith event into multiple events might be a good idea.

Reply:
I agree how the current monolith event should be split into multiple events so I looked at the notification and the response object provided in the [documentation](https://developers.google.com/youtube/v3/guides/push_notifications). Unfortunately, I don't think it's possible to divide the notification into three different types of events. I think you can definitely expand this bee for more functionalities such as uploading video, but in the aspect of notification this is as specific as it can get I believe.

Reply:
I'm not sure if the publication/updated date are initially set to the same value for initial publication, but if they were you could check if they were the same, and at least divide them into two - the initial publish and the update.

Reply:
oof good catch. I missed that. I'll commit that this weekend!